### DISCUSSION TITLE
About --recover option

### DISCUSSION THREAD

**Discussion by ykvishal** [idaholab/moose#17595] [Q&A General] [answered]
I am having 2 issues with --recovery option:

1) --recover option works nicely from Time-step=2  (ie recovering and restarting simulation with Time-step=1 data) onward. But, it does not work for Time-step=1 (ie restarting simulation with Time-step=0 data).  It gives me following error:
> *** ERROR ***
> Internal error in Transient executioner: _t_step is equal to 0 while recovering in init().

I think that this is coming from [Transient.C](https://github.com/idaholab/moose/blob/06bc1f73f0d4dfb3d01ce9ba77d6bb97346e07eb/framework/src/executioners/Transient.C)
```
if (_app.isRecovering()) // Recover case
  {
    if (_t_step == 0)
      mooseError("Internal error in Transient executioner: _t_step is equal to 0 while recovering "
                 "in init().");

    _dt_old = _dt;
  }
```
For my large scale grain growth simulations, it takes several hours (with 560 cores) to reach to Time-step=0. I am hoping that there will be some solution to this issue. Otherwise, I will to have repeat everything again for several hours. I tried to use --restart but it repeats most of the steps from initialization.

2) I am using Nemesis to write output after every 20 time-steps. Due to some [issue](https://github.com/idaholab/moose/discussions/17422), I need to restart simulation many times and I use --recover option to restart simulation. And, nemesis does not append the previous nemesis output files and writes new output over old data. I don't have any issue with csv output. 

If you have any suggestions then please let me know. Thank you.

**Comment by GiudGiud** [comment, marked as answer]
Hello

So you are trying to recover after just the initialization? Or after the first solve?

Regarding your second issue, I m not sure I understand it very well.
If you want to have a nemesis file for each restart attempt, you can use the append_date option to keep the old ones.
If you want to examine multiple time steps in a single file, nemesis files should already contain the outputs at each time step. 

Guillaume

**Reply by ykvishal** [reply]
Hi Guillaume,

Thank you for your response.

`So you are trying to recover after just the initialization? `

Yes, just after initialization. I am trying to run a 3D grain growth simulations with 10K-20K grains. I managed to run a 3D simulations with 6K grains. But, data does not look good as mesh is not that refined. So I increased mesh size and I am also using adaptivity option. It takes almost 28 hrs to reach to initialization step and it fails for Time-step =1 calculation due to memory issue. So I want to use recover option for same number of cores and increase memory using `pmem` option. However, this is not possible with --recover option. I have to go through 28 hrs of initialization step every time if I change something.  


`If you want to examine multiple time steps in a single file, nemesis files should already contain the outputs at each time step.`

Yes, I want to do write and append after every 20 interval. But, somehow nemesis with --recover option rewrites on previous output files. Without --recover option, I don't have this issue. My nemesis block is like this

>   [./nemesis]
>     type = Nemesis
>     interval = 20
>   [../]
> 

Thank you very much.

Vishal

**Reply by GiudGiud** [reply]
Ok thanks for making these two issues clear.
I don't think anyone is planning to add these features to --recover right now. You can create a feature request as a separate github Issue for each.

I suspect both would not be too much work if you want to add the capability yourself. We would gladly welcome a PR.

**Reply by ykvishal** [reply]
Thank you  @GiudGiud . Following you suggestions, I already fixed 1st problem and I figured out solution for 2nd issue. I will do more testing and if everything works then I will also follow your last suggestion.
